src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -511,14 +511,13 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         };
         let msg = format!("use parentheses to call the {}", callable);
 
-        let obligation = self.mk_obligation_for_def_id(
-            trait_ref.def_id(),
-            output_ty.skip_binder(),
-            obligation.cause.clone(),
+        let new_obligation = self.mk_trait_obligation_with_new_self_ty(
             obligation.param_env,
+            trait_ref,
+            output_ty.skip_binder(),
```

Comment:
Is there a reason that we think `output_ty` cannot reference bound variables? (The assertion suggests we do, but I don't immediately see it.)

I'd prefer to see `output_ty.no_bound_vars().unwrap()` if we believe this to be true ([`no_bound_vars`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Binder.html#method.no_bound_vars) docs)

Reply:
I think a function return type can indeed have late bound variables. Perhaps we should just bail out if it does? Alternatively, is there a way to preserve binding levels within `mk_trait_obligation_with_new_self_ty`?